packages/babel-plugin-proposal-destructuring-private/src/util.ts

```diff
@@ -0,0 +1,470 @@
+      element = element.argument as t.Identifier | t.Pattern;
+    } else {
+      newElements.push(tempId);
+    }
+    if (element.type === "AssignmentPattern") {
+      transformed.push({
+        left: element.left,
+        right: transformAssignmentPattern(element.right, tempId),
+      });
+    } else {
+      transformed.push({
+        left: element,
+        right: cloneNode(tempId),
+      });
+    }
+  }
+  return { elements: newElements, transformed };
+}
+
+/**
+ * A DFS simplified pattern traverser. It skips computed property keys and assignment pattern
+ * initializers. The following nodes will be delegated to the visitor:
+ * - ArrayPattern
+ * - ArrayPattern elements
+ * - AssignmentPattern
+ * - ObjectPattern
+ * - ObjectProperty
+ * - RestElement
+ * @param root
+ * @param visitor
+ */
+export function* traversePattern(
+  root: t.LVal,
+  visitor: (
+    node: t.LVal | t.ObjectProperty,
+    index: number,
+    depth: number,
+  ) => Generator<any, void, any>,
+) {
+  const stack = [];
+  stack.push({ node: root, index: 0, depth: 0 });
+  let item: {
+    node: t.LVal | t.ObjectProperty | null;
+    index: number;
+    depth: number;
+  };
+  while ((item = stack.pop()) !== undefined) {
+    const { node, index } = item;
+    if (node === null) continue;
+    yield* visitor(node, index, item.depth);
+    const depth = item.depth + 1;
+    switch (node.type) {
+      case "AssignmentPattern":
+        stack.push({ node: node.left, index: 0, depth });
+        break;
+      case "ObjectProperty":
+        // inherit the depth and index as an object property can not be an LHS without object pattern
+        stack.push({ node: node.value, index, depth: item.depth });
+        break;
+      case "RestElement":
+        stack.push({ node: node.argument, index: 0, depth });
+        break;
+      case "ObjectPattern":
+        for (let list = node.properties, i = list.length - 1; i >= 0; i--) {
+          stack.push({ node: list[i], index: i, depth });
+        }
+        break;
+      case "ArrayPattern":
+        for (let list = node.elements, i = list.length - 1; i >= 0; i--) {
+          stack.push({ node: list[i], index: i, depth });
+        }
+        break;
+      case "TSParameterProperty":
+      case "TSAsExpression":
+      case "TSTypeAssertion":
+      case "TSNonNullExpression":
+        throw new Error(
+          `TypeScript features must first be transformed by ` +
+            `@babel/plugin-transform-typescript.\n` +
+            `If you have already enabled that plugin (or '@babel/preset-typescript'), make sure ` +
+            `that it runs before @babel/plugin-proposal-destructuring-private.`,
+        );
+      default:
+        break;
+    }
+  }
+}
+
+export function hasPrivateKeys(pattern: t.LVal) {
+  let result = false;
+  traversePattern(pattern, function* (node) {
+    if (isObjectProperty(node) && isPrivateName(node.key)) {
+      result = true;
+      yield;
```

Comment:
I believe this is now just preventing `traversePattern`'s `while` loop from continuing. Can you add a comment?